Fix DependencyCache.fetchSourcesFor() performance regression and incorrect sources jar search logic.

artem-zinnatullin opened the pull request:
Fixes #473 caused by performance and logic regression in `DependencyCache.fetchSourcesFor()` introduced in PR #423.

This fix reduces `:setupOkbuck` from 1m+ to ~6 seconds in our project 😸 🚀  🎉  

---

### Detailed explanation of the issue

#### Performance regression:

When some dependency is stored as `.jar` in near to top-level of the project directory, i.e `project/libs` (which is bad but sometimes that happens), then OkBuck will do `fileTree` lookup **in the project directory** for each such dependency. 

In our case we had 4 jars stored in the project directory and fileTree lookup for sources jar took about 15-25 seconds each (lots of files in `.git`, `.okbuck`, `build`, etc folders) which **resulted in 60 seconds+ execution time** of `:setupOkbuck` task.

#### Logic regression:

OkBuck could find sources jar from wrong library by full `fileTree` lookup in the project directory since most of the jars are named only as `artifactId` and do not contain `groupId` in them. See http://blog.joda.org/2017/04/java-se-9-jpms-module-naming.html (Ownership and Uniqueness).

i.e.:

```console
module-1/libs/spark.core-1.0.jar         # Spark Framework
module-2/libs/spark.core-1.0.jar         # Apache Spark
module-2/libs/spark.core-1.0-sources.jar # Apache Spark
```

OkBuck will use `Apache Spark` sources for `Spark Framework` jar.

---

### Detailed explanation of the fix in this PR

If jar is in the project directory — then try to find sources jar only in the same directory as the jar itself, which avoids both expensive `fileTree` lookup and logic regression.

Otherwise use existing strategy to find sources jar in `jar/../../` directory (works for Gradle/Maven cache).

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/uber/okbuck?pullRequest=479) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/uber/okbuck?pullRequest=479) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/uber/okbuck?pullRequest=479) it.</sub>

kageiit commented on buildSrc/src/main/groovy/com/uber/okbuck/core/dependency/DependencyCache.groovy:
```diff
@@ -109,13 +110,21 @@ class DependencyCache {
-                } else if (sourceJars.size() > 1) {
-                    throw new IllegalStateException("Found multiple source jars: ${sourceJars} for ${dependency}")
+                if (FileUtils.directoryContains(rootProject.projectDir, dependency.depFile)) {
+                    // Jas is in the project directory, try to use sources jar right next to the jar itself.
```
Typoe - > Jar

artem-zinnatullin replied:
Typoe 😹 

![hahaha-classic-3602179](https://user-images.githubusercontent.com/967132/28442873-7bf13374-6d67-11e7-8fc6-9d3519dba4aa.png)

kageiit commented on buildSrc/src/main/groovy/com/uber/okbuck/core/dependency/DependencyCache.groovy:
```diff
@@ -109,13 +110,21 @@ class DependencyCache {
-                    sourcesJar = sourceJars[0]
-                } else if (sourceJars.size() > 1) {
-                    throw new IllegalStateException("Found multiple source jars: ${sourceJars} for ${dependency}")
+                if (FileUtils.directoryContains(rootProject.projectDir, dependency.depFile)) {
```
Can you check if dependency is local first? There is a property `depenency.isLocal` and this logic should only apply to such dependencies

artem-zinnatullin replied:
Oh, that's actually better!

Looks like it's okay to actually replace `FileUtils.directoryContains()` with `dependency.isLocal`, please correct me if I'm wrong :)

kageiit commented:
As an integration test, can you add sources to the existing local dependency in the okbuck repo? so we can see it is being picked up correctly

artem-zinnatullin commented:
The only local dependency I've found in OkBuck is `app/libs/buck-android-support.jar`. 

Do you want me to put fake/real sources jar next to it or put some other example jar with sources next in that directory?

artem-zinnatullin commented:
I'm also not sure what do you mean by integration test.

Check if sources are visible in IntelliJ OR create automated test that'll run `./buckw project` and check that `.okbuck/caches` contains both `.jar` and `-sources.jar`?

kageiit commented:
We have a local aar in okbuck
 Travis build everything already.

I'm suggesting you add a sources.jar file right next to the local aar so when Travis runs the tests, it gets picked up

artem-zinnatullin commented:
Alright, added sources jar for rxscreenshotdetector and also added version of it to the filename to ease release tracking.

`./buckw project` finds and copies `-sources.jar` to the `.okbuck/cache`!

artem-zinnatullin commented:
Can you please ship v0.23.1 with this fix?

kageiit commented:
v0.23.1 is live

artem-zinnatullin commented:
Thanks, appreciate that! ❤️
